engine/paymail/paymail_service_client.go

```diff
@@ -0,0 +1,209 @@
+}
+
+type service struct {
+	cs cachestore.ClientInterface
+	pc paymail.ClientInterface
+}
+
+// NewServiceClient creates a new paymail service client
+func NewServiceClient(cs cachestore.ClientInterface, pc paymail.ClientInterface) (ServiceClient, error) {
+	if pc == nil {
+		return nil, spverrors.Newf("paymail client is required")
+	}
+	return &service{
+		cs: cs,
+		pc: pc,
+	}, nil
+}
+
+// GetSanitizedPaymail validates and returns the sanitized version of paymail address (alias@domain.tld)
+func (s *service) GetSanitizedPaymail(addr string) (*paymail.SanitisedPaymail, error) {
+	if err := paymail.ValidatePaymail(addr); err != nil {
+		return nil, err //nolint:wrapcheck // we have handler for paymail errors
+	}
+
+	sanitized := &paymail.SanitisedPaymail{}
+	sanitized.Alias, sanitized.Domain, sanitized.Address = paymail.SanitizePaymail(addr)
+
+	return sanitized, nil
+}
+
+// GetCapabilities is a utility function to retrieve capabilities for a Paymail provider
+func (s *service) GetCapabilities(ctx context.Context, domain string) (*paymail.CapabilitiesPayload, error) {
+	// Attempt to get from cachestore
+	// todo: allow user to configure the time that they want to cache the capabilities (if they want to cache or not)
+	capabilities := new(paymail.CapabilitiesPayload)
+	if s.cs != nil {
+		if err := s.cs.GetModel(
+			ctx, cacheKeyCapabilities+domain, capabilities,
+		); err != nil && !errors.Is(err, cachestore.ErrKeyNotFound) {
+			return nil, spverrors.Wrapf(err, "failed to get capabilities from cachestore")
+		} else if len(capabilities.Capabilities) > 0 {
+			return capabilities, nil
+		}
+	}
+
+	// Get SRV record (domain can be different!)
+	var response *paymail.CapabilitiesResponse
+	srv, err := s.pc.GetSRVRecord(
+		paymail.DefaultServiceName, paymail.DefaultProtocol, domain,
+	)
+	if err != nil {
+		// Error returned was a real error
+		if !strings.Contains(err.Error(), "zero SRV records found") { // This error is from no SRV record being found
+			return nil, err //nolint:wrapcheck // we have handler for paymail errors
+		}
+
+		// Try to get capabilities without the SRV record
+		// 'Should no record be returned, a paymail client should assume a host of <domain>.<tld> and a port of 443.'
+		// http://bsvalias.org/02-01-host-discovery.html
+
+		// Get the capabilities via target
+		if response, err = s.pc.GetCapabilities(
+			domain, paymail.DefaultPort,
+		); err != nil {
+			return nil, err //nolint:wrapcheck // we have handler for paymail errors
+		}
+	} else {
+		// Get the capabilities via SRV record
+		if response, err = s.pc.GetCapabilities(
+			srv.Target, int(srv.Port),
+		); err != nil {
+			return nil, err //nolint:wrapcheck // we have handler for paymail errors
+		}
+	}
+
+	// Save to cachestore
+	if s.cs != nil && !s.cs.Engine().IsEmpty() {
+		_ = s.cs.SetModel(
```

Comment:
I would suggest to at least log the error - but, I guess, here the logger is not available 🤔